proposals/scalable-robust-operator.md

```diff
@@ -0,0 +1,284 @@
+    stress on the apiserver as the number of launchers increases.
+  - The launcher pod has execution permissions on any other Pod in the
+    namespace, which can be a security concern.
+  - The v1 controller doesn’t implement launcher pod retries, although there are
+    plans to. So the MPIJob behaves like a plain Pod in this version.
+
+## Design
+
+In order to address the problems of the existing architecture, we propose the
+following changes:
+
+- **The use of `ssh` instead of `kubectl exec`.**
+  
+  This would avoid any pod-to-pod communication happening through apiserver and
+  doesn’t require giving execution permissions at the namespace level.
+  This can be achieved like this:
+  - The controller can generate a single key and share it with the launcher and
+    worker pods through a Secret.
+  - The launcher and workers mount the Secret and set appropriate file
+    permissions.
+  - The workers run an SSH server instead of `sleep`.
+  - When encrypted communication is not a requirement, users could have the
+    choice to use `rsh` for faster communication.
+
+- **The use of stable hostnames and a headless Service for the workers**
+  - This removes the need to query Pod IPs, as the Pods can discover each other
+    through DNS resolution. The hostfile can be generated statically by the
+    controller using the stable hostnames.
+  - Starting with k8s 1.22, we can use
+    [Indexed Jobs with stable hostnames](https://git.k8s.io/enhancements/keps/sig-apps/2214-indexed-job)
+    to delegate the pod management to Kubernetes. Additionally, this will give
+    us robust failure tracking so that we can give users control over retry
+    limits.
+    In the meantime, we can continue using plain Pods.
+    - Caveat 1: The Job controller doesn’t report the number of running Pods.
+      Instead, it reports active Pods, which include running and pending
+      (scheduling or starting). But Kubernetes SIG Apps is
+      [open to add a status field for running Pods](https://kubernetes.slack.com/archives/C18NZM5K9/p1619549430067400).
+    - Caveat 2: Horovod supports elastic workers, but the Kubernetes Job
+      doesn’t support changes to the completions field. This can be supported
+      starting from 1.23. In the meantime, we can replicate the behavior by
+      creating a new Job and doing Pod adoption.
+  - For Intel MPI, we also need a headless Service to front the launcher.
+- **Revert the use of the Job API for the launcher.**
+  - The Job controller handles retries when the launcher or any of the workers fail.
+  - Caveat 1 also applies: The Job controller doesn’t report if the Pod is running.
+    We can continue watching Pods in the meantime.
+- With the above changes, **the following objects can be removed**:
+  - The ServiceAccount+Role+RoleBinding for the launcher.
+  - The `kubectl-delivery` init container in the launcher, as there is no need
+    to obtain IPs, speeding up startup time.
```

Comment:
Do u mean the init-container `kubectl-delivery` will be removed? If that, how could we  keep the launcher start after the  workers?

Reply:
There are 2 options:
1. Do nothing, leave such handling to the gang scheduler. For this reason it is important that the launcher can do retries.
1. To not create the launcher until all the worker pods are running. So everything is handled by the controller, no need for extra cache syncs.

Reply:
I don't think so, you would like to speed up startup time, but it does not if you want the launcher to retry for failures. And the scheduler should not know about the job startup strategy.

Reply:
By the same time, how do you verify the  real launcher failed reason for option 1? for example,
* workers does not startup, the launcher failed and restart
* training job failed, the launcher failed.

Reply:
What do you mean by "job startup strategy"?

As for differentiating failures, I don't think we can or should do that. There is another type of failure which is kind of a mix of the two you described: a worker goes down. In this case, the job might have already started.

But perhaps option 2 is reasonable. WDYT?

Reply:
If we want to support retries, then by product we want option 1, correct?

Reply:
we can still use option 2. It might make startups less flaky. But it's not a guarantee anyways. A worker can simply just fail in between the time it reported running and the launcher started running.